### ISSUE TITLE
Named assignments break disambiguation

### ISSUE THREAD

**Issue by alberth** [igordejanovic/parglare#27]
* parglare version: HEAD
* Python version: 3.5.2
* Operating System: linux ubuntu

### Description

Having fun trying to shoehorn a line-based language into the parser, which is somewhat asking for trouble of course. The best solution so far is to make `LAYOUT` handle only spaces and tabs, and have a dedicated `eol` non-terminal that handles line endings. That seems to mostly work except the grammar needs to handle truly empty lines explicitly (not added here for simplicity).

Anyway, while experimenting I ran into a weird disambiguation problem. See
``` python
# --------------------
gram_text = r"""
typedefheader : "define" "type"  eol ;

eol : /\n/
| filecomment
| doccomment
;

filecomment : /#.*\n/ ;
doccomment : /#<.*\n/ {20} ;

LAYOUT: EMPTY | /[ \t]+/ ;
KEYWORD: /[-a-z]+/ ;
"""
# --------------------
text = "define type #< something\n"

g = Grammar.from_string(gram_text)
p = Parser(g)
print(p.parse(text))
print("############################")
print()

# --------------------
gram_text2 = r"""
typedefheader : "define" "type"  eol ;

eol : /\n/
| filecomment
| doccomment
;

filecomment : /#.*\n/ ;
doccomment : docu=/#<.*\n/ {20} ;    // Added "docu="

LAYOUT: EMPTY | /[ \t]+/ ;
KEYWORD: /[-a-z]+/ ;
"""
# --------------------

g2 = Grammar.from_string(gram_text2)
p2 = Parser(g2)
print(p2.parse(text))
```
In the first grammar, the `eol` defines a `\n` possibly prefixed by two different forms of comment, `# filecomment` and `#< documentation comment`. Since these two forms of comment are obviously ambiguous, the second form has a nice `{20}` attached to it to give it priority. So far so good.

Now I wanted to know if it actually worked, so I added `docu=` to the `doccomment` rule (printing the result would give me a named instance rather than plain text). However, the parser then crashed on the ambiguity:
``` python console
['define', 'type', '#< something\n']
############################

Traceback (most recent call last):
  File "/home/hat/projects/psi/psi5/compiler3/parglare/parser.py", line 213, in parse
    ntok = next_token(cur_state, input_str, position)
  File "/home/hat/projects/psi/psi5/compiler3/parglare/parser.py", line 537, in _next_token
    ntok = self._lexical_disambiguation(tokens)
  File "/home/hat/projects/psi/psi5/compiler3/parglare/parser.py", line 672, in _lexical_disambiguation
    raise DisambiguationError(tokens)
parglare.exceptions.DisambiguationError: [<filecomment(#< something
)>, <#<.*\n(#< something
)>]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "named_assignment.py", line 43, in <module>
    print(p2.parse(text))
  File "/home/hat/projects/psi/psi5/compiler3/parglare/parser.py", line 217, in parse
    disambiguation_error(e.tokens))
parglare.exceptions.ParseError: Error at position 1,12 => "fine type *#< somethi". Can't disambiguate between: <filecomment(#< something
)> or <#<.*\n(#< something
)>
```
Note the second line with hashes in the output, meaning that the first parse was fine.

**Comment by igordejanovic** [comment]
This is indeed strange. Parsing behaviour shouldn't depend on the assignments at all.
Thanks for the report.

**Comment by igordejanovic** [comment]
I just realized what was the problem. Adding `docu=` to `doccomment` will make the rule non-terminal e.g. there will be implicit regex match `/#<.*\n/` with the default priority and the priority of the `doccomment` production will be set to `20`. What you should do is to move your regex match to a terminal rule with raised priority:

```
doccomment : docu=doccomment_match;
doccomment_match: /#<.*\n/ {20};
```
So, this is expected behaviour after all :)

**Comment by alberth** [comment]
It might be expected behaviour for you, but not for me. The documentation on assigned names says nothing  about side effects like introducing new intermediate rules that I didn't write, or breaking priorities. 
Priorities documentation says nothing about how priorities don't work if they are at different levels. (Although I would never look for that until I realize that a new level is introduced.)

To me, I wrote in `doccomment: /#<.*\n/ {20};` what also your first intuition was, a single terminal rule with priority, where after matching, a little extra feature of adding a named result is added. I don't know the implementation at all.

**Comment by igordejanovic** [comment]
Right. Docs needs a bit more explanation in this part. The problem was that syntaxes for defining terminal rule and non-terminal rule are similar. If you use string match or regex match in your non-terminal rule parglare will create terminal rule for you. You can override that of course and provide your terminal definition but that is required only if you want to change terminal priority or use some other disambiguation on the terminal. Terminal rules don't allow assignments as it doesn't make much sense in my view. They should be a simple types/strings, not objects with attributes.

If you have an idea where and how to extend the docs to make this more clear please advise.

**Comment by alberth** [comment]
Reading this I am not convinced the docs are the first thing that needs consideration.

In my view, languages should have concepts that are orthogonal to each other. For every concept, it should work everywhere, and it should always work in the same way, no matter what other concepts are used around it.
Unfortunately, in real life, concepts are often not entirely orthogonal, for various reasons. In that case, I prefer explicitness. I like the tool to complain about my impossible wish, rather than changing the semantics. I would be forced to explicitly write the two rules that you proposed as solution.

The disadvantage of this approach is that a user hits the limits of the tool a little earlier, but it's a clean limit. Either the input is accepted and each concept works as advertised or it fails.
You can push the boundaries a little further like you do in the combined named-assignment-priorities-terminal-rule case, but I got confused, and reported it as a bug. Right now, you're considering how to explain this all in the documentation. In my view, it's all muddy water of edge cases with weird highly complicated semantics depending on the exact combination of concepts that only you can truly grasp.

Things I would consider:
- Implement named assignments for terminal rules. They may not make much sense, but if it's easy to support, and a user sees a use for it apparently, why not?
- Forbid the combined named assignments and priority for terminal rules. They make no sense, don't try that! Simplest is forbid feature A if the feature B is also present, so in the docs, you can explain how to do the combination near the forbidden feature A.

**Comment by igordejanovic** [comment]
After reading this and giving it some thought I think you are absolutely right. Much of the complexity/confusion would go away if terminals are defined e.g. in a dedicated block of grammar code. 

The problem is not that the combined use of assignment and priority is forbidden for terminals but assignment alone is. So using assignment makes a rule non-terminal. You got confused as your terminal rule suddenly become non-terminal without you knowing it. Forcing a user to specify his terminals in the dedicated block would prevent these sort of errors as this block will only allow a proper terminal definitions.

Still I find cumbersome for a user to be forced to define each terminal he uses (especially for string matches) but making it clear where terminal should be defined in case one needs to change disambiguation rules will make it all much cleaner. I usually tend to write all terminal rules at the end of the grammar anyway :)

Thanks for the fruitful discussion and ideas. I'm registering a new issue that should deal with changing the way terminal rules are defined.

**Comment by alberth** [comment]
A separate block is one option, and probably preferable to having to specify it with each rule.

You can have it implicitly, but you have to make clear how it works in term of the language concepts. Define in the documentation explicitly when a rule is a terminal rule, and define all other rules as non-terminal rules (if you have only 2 kinds of rules). I don't remember seeing this, but I may have missed it. With assignment say it only works for non-terminal rules. If a user tries it anyway, point to "only allowed in a non-terminal rule" so the user hopefully finds the discussion.

Having a terminal rule (or even a terminal section), doesn't mean you cannot have terminal literals in non-terminal rules. Simplest is perhaps to define them as a rewrite to an unused terminal rule name that gets added to the terminal section. (The implementation may not actually do that, but it's an implementation detail, much like your smart LR compression that is fully hidden for me. The implementation does however have to keep the illusion consistent, docs are leading rather than implementation.)

As long as both forms are fully equivalent, a user is free to pick whatever he/she prefers, until he/she wants to use a feature that is not defined for one kind of rules (at that point a rewrite to the more basic form is required).

**Comment by igordejanovic** [comment]
Development version now requires separate `terminals` section for terminal definition so this issue will become obsolete. This will be released shortly in v0.6.0.
